File: src/main/java/com/baulsupp/oksocial/dns/DnsOverride.java
package com.baulsupp.oksocial.dns;

import com.baulsupp.oksocial.UsageException;
import com.google.common.collect.Maps;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import okhttp3.Dns;

public class DnsOverride implements Dns {
  private final Dns dns;
  private Map<String, String> overrides = Maps.newHashMap();

  public DnsOverride(Dns dns) {
    this.dns = dns;
  }

  private void put(String host, String target) {
    overrides.put(host, target);
  }

  @Override public List<InetAddress> lookup(String s) throws UnknownHostException {
    String override = overrides.get(s);

    if (override != null) {
      return Arrays.asList(InetAddress.getByName(override));
    }

    return dns.lookup(s);
  }

  public static DnsOverride build(Dns dns, String resolveString) {
    DnsOverride dnsOverride = new DnsOverride(dns);

    String[] parts = resolveString.split(":");

    if (parts.length != 2) {
      throw new UsageException("Invalid resolve string '" + resolveString + "'");
    }

    dnsOverride.put(parts[0], parts[1]);

    return dnsOverride;
  }
}

File: src/main/kotlin/com/baulsupp/okurl/network/DnsOverride.kt
package com.baulsupp.okurl.network

import com.baulsupp.schoutput.UsageException
import okhttp3.Dns
import java.net.InetAddress
import java.util.logging.Logger

class DnsOverride(private val dns: Dns) : Dns {
  private val overrides = linkedMapOf<String, String>()

  private fun put(host: String, target: String) {
    overrides[host] = target
  }

  override fun lookup(hostname: String): List<InetAddress> {
    val override = overrides[hostname]

    if (override != null) {
      logger.fine("Using Dns Override ($hostname): $override")
      return listOf(InetAddress.getByName(override))
    }

    return dns.lookup(hostname)
  }

  companion object {
    private val logger = Logger.getLogger(DnsOverride::class.java.name)

    fun build(dns: Dns, resolveStrings: List<String>): DnsOverride {
      val dnsOverride = DnsOverride(dns)

      for (resolveString in resolveStrings) {
        val parts = resolveString.trim().split(":", limit=2)

        if (parts.size != 2) {
          throw UsageException("Invalid resolve string '$resolveString'")
        }

        dnsOverride.put(parts[0], parts[1])
      }

      return dnsOverride
    }
  }
}
